pkg/services/cosmosdb/cosmos-utils.go

```diff
@@ -178,3 +182,218 @@ func deleteDatabase(
 	}
 	return nil
 }
+
+// The deployment will return success once the write region is created,
+// ignoring the status of read regions , so we must implement detection logic
+// by ourselves.
+func (c *cosmosAccountManager) waitForReadLocationsReady(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	dt := instance.Details.(*cosmosdbInstanceDetails)
+	resourceGroupName := instance.ProvisioningParameters.GetString("resourceGroup")
+	accountName := dt.DatabaseAccountName
+	databaseAccountClient := c.databaseAccountsClient
+
+	err := pollingUntilReadLocationsReady(
+		ctx,
+		resourceGroupName,
+		accountName,
+		databaseAccountClient,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return dt, nil
+}
+
+// For sqlAllInOneManager, the real type of `instance.Details` is
+// `*sqlAllInOneInstanceDetails`, so type assertion must be changed.
+// Expect type assertion, this function is totally the same as previous one.
+func (s *sqlAllInOneManager) waitForReadLocationsReady(
+	ctx context.Context,
+	instance service.Instance,
+) (service.InstanceDetails, error) {
+	dt := instance.Details.(*sqlAllInOneInstanceDetails)
+	resourceGroupName := instance.ProvisioningParameters.GetString("resourceGroup")
+	accountName := dt.DatabaseAccountName
+	databaseAccountClient := s.databaseAccountsClient
+
+	err := pollingUntilReadLocationsReady(
+		ctx,
+		resourceGroupName,
+		accountName,
+		databaseAccountClient,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return dt, nil
+}
+
+const succeeded = "succeeded"
+
+// For now, this method will return on either context is cancelled or
+// every region's state is "succeeded" in seven consecutive check.
+// The reason why we need seven consecutive check is that
+// the read region is created one by one, there is a small gap between
+// the finishment of previous creation and the start of the next creation.
+// By this check, we can detect gaps shorter than 1 mintue,
+// and report success within 70 seconds after completion.
+func pollingUntilReadLocationsReady(
+	ctx context.Context,
+	resourceGroupName string,
+	accountName string,
+	databaseAccountClient cosmosSDK.DatabaseAccountsClient,
+) error {
+	childCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Seven comsecutive check is needed to report success.
+	const confirmNumberOfTimes = 7
+
+	ticker := time.NewTicker(time.Second * 10)
+	previousSucceededTimes := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			result, err := databaseAccountClient.Get(
+				childCtx,
+				resourceGroupName,
+				accountName,
+			)
+			if err != nil {
+				return err
+			}
+
+			//Check whether every read location's state is "Succeeded"
+			allSucceed := true
+			readLocations := *(result.DatabaseAccountProperties.ReadLocations)
```

Comment:
we do something similar when we check for parent/child relationships with the async engine. I'd prefer if we used that here as well, vs polling and blocking in this.

Reply:
Seems this case is different from checking for parent/child relationships. Checking for parent/child relationships is a general operation and it is done before the first provision step is executed. It's reasonable to make checking a `delayedTask` and trigger the real executing task after the checking is completed. But here, this function itself will be executed inside an async task. Will it be a little strange if we make it a `delayedTask` and manipulate async engine directly here?

Reply:
Hi @jeremyrickard , maybe you are worry about `polling`? Actually this function works like [WaitForCompletion](https://github.com/Azure/open-service-broker-azure/blob/master/pkg/services/cosmosdb/common_deprovision.go#L44) in deprovision, it is executed in the go routine and won't block any other operation.

Reply:
Hi @jeremyrickard , any follow-up on this topic?

Reply:
@norshtein I'm occupied with other things today so I probably won't get a chance to circle back to this until Monday, my apologies! I'll revisit this first thing in the morning on Monday!

Reply:
@norshtein @jeremyrickard I think the async stuff is being approached reasonably here. `waitForReadLocationsReady` is registered as a provisioning/update step and the general requirements for those is that they be idempotent and interruptable, and those conditions seem to be met here. `waitForReadLocationsReady` calls `pollingUntilReadLocationsReady` and that is interruptable and it seems that if the the application process died while that was executing, it would seamlessly be resumed by the async engine (like any other provisioning step) and would eventually succeed.

I do have other issues with the implementation here, but this isn't one of them. I wouldn't worry about this particular detail. I'll comment on the other issues closer to where they appear in the code...

Reply:
Hi @krancour , if we change the implementation of `waitForReadLocationsReady` as discussed below, the function itself will be idempotent and interruptable. If the process died for some reason, when it is restarted, it is still able to report success when all regions are created.

Besides, I'm a little confused how does the async engine deals with the process termination. I'm guessing that when the application is restarted, the async engine will find tasks in the task queue and rerun them. If I guessed it right, then the previous implementation should also be idempotent and interruptable. `waitForReadLocationsReady` will be rerun, invoking `pollingUntilReadLocationsReady` function, waiting until all read locations ready. Or my guess is totally wrong and the async engine doesn't work in this way...

Reply:
@norshtein You're correct the current implementation would be. My apologies for the confusion.